### PR TITLE
net/devif: allocate devif callback dynamically

### PR DESCRIPTION
**Pull request by anchao**
## Summary

net/devif_callback: add support for CONFIG_NET_ALLOC_CONNS

## Impact

netdev interface callback allocate

## Testing

iperf client/server test

### CODE REVIEW

File: net/devif/devif_callback.c [hunk 1]
```diff
@@ -262,7 +276,7 @@ FAR struct devif_callback_s *
   /* Check the head of the free list */
 
   net_lock();
-  ret  = g_cbfreelist;
+  ret = devif_callback_alloc_internal();
```
Comment (xiaoxiang781216) [inline net/devif/devif_callback.c:279]: let's merge devif_callback_alloc_internal into devif_callback_alloc?

File: net/devif/devif_callback.c [hunk 2]
```diff
@@ -35,14 +35,15 @@
 #include <nuttx/net/net.h>
 #include <nuttx/net/netdev.h>
 
+#include <nuttx/kmalloc.h>
```
Comment (xiaoxiang781216) [inline net/devif/devif_callback.c:38]: move before line 34

File: net/socket/Kconfig [hunk 3]
```diff
@@ -5,13 +5,12 @@
...
-	int "Max socket operations"
-	default 16 if !DEFAULT_SMALL
-	default 4 if DEFAULT_SMALL
+config NET_NACTIVESOCKETS_PER_BLOCK
+	int "Max socket operations per block"
+	default 4
 	---help---
-		Maximum number of concurrent socket operations (recv, send,
-		connection monitoring, etc.). Default: 16
+		Maximum number of concurrent socket operations per block
```
Comment (xiaoxiang781216) [inline net/socket/Kconfig:12]: The number of concurrent socket operations per allocation

File: net/socket/Kconfig [hunk 4]
```diff
@@ -5,13 +5,12 @@
 
 menu "Socket Support"
 
-config NET_NACTIVESOCKETS
-	int "Max socket operations"
-	default 16 if !DEFAULT_SMALL
-	default 4 if DEFAULT_SMALL
+config NET_NACTIVESOCKETS_PER_BLOCK
+	int "Max socket operations per block"
+	default 4
```
Comment (xiaoxiang781216) [inline net/socket/Kconfig:10]: default to 1 if DEFAULT_SMALL

**Review by Ouss4 — COMMENTED** [review]
Since now we are allocating the callback structure dynamically, shouldn't `devif_callback_free` also free this buffer?

File: net/devif/devif_callback.c (hunk 1 above)
Comment (anchao) [reply net/devif/devif_callback.c]: Done

File: net/devif/devif_callback.c (hunk 2 above)
Comment (anchao) [reply net/devif/devif_callback.c]: Done

File: net/socket/Kconfig (hunk 3 above)
Comment (anchao) [reply net/socket/Kconfig]: Done

File: net/socket/Kconfig (hunk 4 above)
Comment (anchao) [reply net/socket/Kconfig:11]: Done

**Comment by anchao** [comment]
> Since now we are allocating the callback structure dynamically, shouldn't `devif_callback_free` also free this buffer?
In order to avoid frequent access to the allocator, the current implement just re-queue the callback entry back to the freelist，we will reuse this area to avoid memory fragmentation

**Comment by anchao** [comment]
> Looking at the implementation of `devif_callback_alloc` right now, it seems to be allocating the whole pool for each call. How is this working?

Yes, each request size is pool based(NET_NACTIVESOCKETS_PER_ALLOC), 
if the freelist is empty, we will allocate for a sequence of callback entries and queue them into the freelist, this is to reduce the memory fragmentation and allocator access cycles.

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by Ouss4** [comment]
> > Looking at the implementation of `devif_callback_alloc` right now, it seems to be allocating the whole pool for each call. How is this working?
> 
> Yes, each request size is pool based(NET_NACTIVESOCKETS_PER_ALLOC), if the freelist is empty, we will allocate for a sequence of callback entries and queue them into the freelist, this is to reduce the memory fragmentation and allocator access cycles.

I see.  Thanks for the explanation.  I thought the allocation was based on an element not a pool and I was going to ask about the impact of this long lived allocation on fragmentation.  But it looks like you've already considered it.

File: net/devif/devif_callback.c [hunk 5]
```diff
@@ -258,11 +235,31 @@ FAR struct devif_callback_s *
...
   FAR struct devif_callback_s *ret;
+  int i;
 
   /* Check the head of the free list */
 
   net_lock();
-  ret  = g_cbfreelist;
+  ret = g_cbfreelist;
+
+  /* Allocate the callback entry from heap */
```
Comment (Ouss4) [inline net/devif/devif_callback.c:245]: @anchao can you please add part of https://github.com/apache/incubator-nuttx/pull/5110#issuecomment-1002741733 as an explanation here?

Comment (anchao) [reply net/devif/devif_callback.c:245]: Of course, let us wait for the conclusion of [#5104](https://github.com/apache/incubator-nuttx/pull/5104), I think this pull request needs to be further enhanced

Comment (xiaoxiang781216) [reply net/devif/devif_callback.c:245]: Let's apply the similar change as #5104

Comment (anchao) [reply net/devif/devif_callback.c:272]: Done

**Review by Ouss4 — APPROVED** [review]

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**